16_remote_allowlist_retrieval.md

```diff
@@ -0,0 +1,256 @@
+    be secure (HTTPS), and non-redirecting (no 30X HTTP codes) to reduce
+    the chance of any MITM attacks. If this option is specified then
+    you can't also have the `--whitelist` option present.
+
+  * `allowlist-sig-url`
+    An HTTP URL that points to the cryptographic signature of the JSON
+    data file. Again, This URL must be secure (HTTPS), and non-redirecting
+    (no 30X HTTP codes) to reduce the chance of any MITM attacks. This
+    option cannot be specified if `allowlist-url` isn't also present. And
+    you must include either this option or `allowlist-sig` or `allowlist-checksum`
+
+  * `allowlist-sig`
+    The cryptographic signature of the remote JSON data file. This
+    option This option cannot be specified if `allowlist-url`
+    isn't also present. And you must include either this option or
+    `allowlist-sig-url` or `allowlist-checksum`. This option gives more
+    flexibility to the user if the `allowlist-sig-url` doesn't quite
+    fit their workflow.
+
+  * `allowlist-key`
+    The key used to verify the `allowlist-sig`. This option is required if
+    `allowlist-sig` or `allowlist-sig-url` is used.
+
+  * `allowlist-checksum`
+    The SHA256 checksum of the allow-list. This can be used to verify
+    that the allow-list hasn't been tampered. You must either verify
+    the checksum or signature (or both) of the allow-list.
+
+* Expand the way allow-lists are passed between the tenant and the
+verifier to use JSON instead of the simple text file that exists now. This
+makes the contract more explicit and allows for expansion in the future.
+
+### User Stories
+
+#### Story 1
+
+As a Fedora CoreOS user, I already know every package that could be
+present on my system based on my ostree version. I want to just tell
+Keylime where to find the list of file hashes for my specific version
+and let it do the rest
+
+#### Story 2
+
+I have a secure build process that can produce allow-lists for the
+software my team is building. I can store these allow-lists in an external
+repository (say an OCI registry). As part of my build/release process I
+have the cryptographic signatures of these allow-lists. I would like to
+tell keylime about these lists when deploying and provide the signature
+to keylime to verify it pulled the expected lists.
+
+### Risks and Mitigations
+
+* We will be updating the internal API between the tenant and the verifier
+to use the new JSON spec. Care will be taken to make sure this is done in
+a backwards compatible way so existing instances of keylime can continue
+to use their existing whitelists.
+
+* We will need to thoroughly review our handling of the external HTTP
+connections to make sure they are as secure as possible and don't do
+things like accept invalid TLS certificates.
+
+* We will need to thoroughly review our handling of the cryptographic
+signatures and checksums
+
+## Design Details
+
+Propsed JSON spec:
+
+{
```

Comment:
There are a few minor mistakes in the example spec and this mistake is on me :smile:.

- 1 missing comma
- 1 extra comma
- Missing `}`
- optional second hash example shouldn't be the same hash function

The JSON should be:

```
{
    "meta": {
        "version": "",        # Hash format version to allow for future changes if needed
        "generator": "",     # What generated the hash payload
        "timestamp": ""     #  When the hash payload was generated
    },
    "release": "",            # Release Identifier. Specifics here may be different for different compound components
    "hashes": {
        "/filesystem/path": [
            {"sha256": "HASH"}, # an object so other hashes could go side by side if needed
            {"sha128": "HASH"}  # other hashes if required/wanted
        ]
    }
}
```